<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Multi Amplicon - a real world example â€¢ MultiAmplicon</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">MultiAmplicon</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/MultiAmplicon-real-world-example.html">Multi Amplicon - a real world example</a>
    </li>
    <li>
      <a href="../articles/MultiAmplicon-small-example.html">MultiAmplicon - a small example</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Multi Amplicon - a real world example</h1>
                        <h4 class="author">Emanuel Heitlinger</h4>
            
            <h4 class="date">2018-04-10</h4>
          </div>

    
    
<div class="contents">
<div id="data-preparation-download" class="section level2">
<h2 class="hasAnchor">
<a href="#data-preparation-download" class="anchor"></a>Data preparation (download)</h2>
<p>We first download the sample data set from SRA using a helper scirtp part of the MultiAmplicon package.</p>
<p>You will need about 2Gb space to store those files if you want to execute the pipeline below on your local computer and/or compile the vignette locally.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SRAdb) 
<span class="co">#&gt; Loading required package: RSQLite</span>
<span class="co">#&gt; Loading required package: graph</span>
<span class="co">#&gt; Loading required package: BiocGenerics</span>
<span class="co">#&gt; Loading required package: parallel</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'BiocGenerics'</span>
<span class="co">#&gt; The following objects are masked from 'package:parallel':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span>
<span class="co">#&gt;     clusterExport, clusterMap, parApply, parCapply, parLapply,</span>
<span class="co">#&gt;     parLapplyLB, parRapply, parSapply, parSapplyLB</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     IQR, mad, sd, var, xtabs</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     anyDuplicated, append, as.data.frame, cbind, colMeans,</span>
<span class="co">#&gt;     colnames, colSums, do.call, duplicated, eval, evalq, Filter,</span>
<span class="co">#&gt;     Find, get, grep, grepl, intersect, is.unsorted, lapply,</span>
<span class="co">#&gt;     lengths, Map, mapply, match, mget, order, paste, pmax,</span>
<span class="co">#&gt;     pmax.int, pmin, pmin.int, Position, rank, rbind, Reduce,</span>
<span class="co">#&gt;     rowMeans, rownames, rowSums, sapply, setdiff, sort, table,</span>
<span class="co">#&gt;     tapply, union, unique, unsplit, which, which.max, which.min</span>
<span class="co">#&gt; Loading required package: RCurl</span>
<span class="co">#&gt; Loading required package: bitops</span>
<span class="co">#&gt; Setting options('download.file.method.GEOquery'='auto')</span>
<span class="co">#&gt; Setting options('GEOquery.inmemory.gpl'=FALSE)</span>
<span class="kw">library</span>(MultiAmplicon)
<span class="co">#&gt; Loading required package: ShortRead</span>
<span class="co">#&gt; Loading required package: BiocParallel</span>
<span class="co">#&gt; Loading required package: Biostrings</span>
<span class="co">#&gt; Loading required package: S4Vectors</span>
<span class="co">#&gt; Loading required package: stats4</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'S4Vectors'</span>
<span class="co">#&gt; The following object is masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     expand.grid</span>
<span class="co">#&gt; Loading required package: IRanges</span>
<span class="co">#&gt; Loading required package: XVector</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'Biostrings'</span>
<span class="co">#&gt; The following object is masked from 'package:graph':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     complement</span>
<span class="co">#&gt; The following object is masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     strsplit</span>
<span class="co">#&gt; Loading required package: Rsamtools</span>
<span class="co">#&gt; Loading required package: GenomeInfoDb</span>
<span class="co">#&gt; Loading required package: GenomicRanges</span>
<span class="co">#&gt; Loading required package: GenomicAlignments</span>
<span class="co">#&gt; Loading required package: SummarizedExperiment</span>
<span class="co">#&gt; Loading required package: Biobase</span>
<span class="co">#&gt; Welcome to Bioconductor</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     Vignettes contain introductory material; view with</span>
<span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class="co">#&gt; Loading required package: DelayedArray</span>
<span class="co">#&gt; Loading required package: matrixStats</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'matrixStats'</span>
<span class="co">#&gt; The following objects are masked from 'package:Biobase':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     anyMissing, rowMedians</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'DelayedArray'</span>
<span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</span>
<span class="co">#&gt; The following object is masked from 'package:Biostrings':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     type</span>
<span class="co">#&gt; The following object is masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     apply</span>
<span class="co">#&gt; Loading required package: dada2</span>
<span class="co">#&gt; Loading required package: Rcpp</span>

if(!<span class="kw">file.exists</span>(<span class="st">"SRAmetadb.sqlite"</span>)){
    sqlfile &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/SRAdb/topics/getSRAdbFile">getSRAdbFile</a></span>()
} else {
    sqlfile &lt;-<span class="st"> "SRAmetadb.sqlite"</span>
}
<span class="co">#&gt; Unzipping...</span>
<span class="co">#&gt; Metadata associate with downloaded file:</span>
<span class="co">#&gt; c("schema version", "creation timestamp")c("1.0", "2018-04-05 23:32:48")</span>

sra_con &lt;-<span class="st"> </span><span class="kw">dbConnect</span>(<span class="kw">SQLite</span>(), sqlfile)

carnivoreSeqRuns &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/SRAdb/topics/getSRA">getSRA</a></span>(<span class="dt">search_terms =</span>
                               <span class="st">'"Intestinal biome sequencing of carnivores"'</span>,
                           <span class="dt">sra_con=</span>sra_con, <span class="dt">acc_only=</span><span class="ot">FALSE</span>)

carnivoreSeqRuns &lt;-<span class="st"> </span>carnivoreSeqRuns[, <span class="kw">c</span>(<span class="st">'run'</span>,<span class="st">'study'</span>,<span class="st">'sample'</span>,
                                         <span class="st">'experiment'</span>, <span class="st">'sample_attribute'</span>)]

runs &lt;-<span class="st"> </span><span class="kw">tail</span>(carnivoreSeqRuns$run, <span class="dt">n=</span><span class="dv">50</span>)

destDir &lt;-<span class="st"> "./download"</span>

if(!<span class="kw">file_test</span>(<span class="st">"-d"</span>, destDir)) <span class="kw">dir.create</span>(destDir)

fastqFiles &lt;-<span class="st"> </span><span class="kw">list.files</span>(destDir, <span class="dt">pattern=</span><span class="st">".fastq.gz"</span>, <span class="dt">full.names=</span><span class="ot">TRUE</span>)

if(<span class="kw">length</span>(fastqFiles) &lt;<span class="st"> </span><span class="dv">100</span>){
    <span class="kw"><a href="http://www.rdocumentation.org/packages/SRAdb/topics/getSRAfile">getSRAfile</a></span>(runs, sra_con, <span class="dt">fileType =</span> <span class="st">'fastq'</span> , <span class="dt">srcType =</span> <span class="st">"ftp"</span>,
               <span class="dt">destDir=</span>destDir)
    fastqFiles &lt;-<span class="st"> </span><span class="kw">list.files</span>(destDir, <span class="dt">pattern=</span><span class="st">".fastq.gz"</span>, <span class="dt">full.names=</span><span class="ot">TRUE</span>)
}</code></pre></div>
</div>
<div id="data-preparation-filtering" class="section level2">
<h2 class="hasAnchor">
<a href="#data-preparation-filtering" class="anchor"></a>Data preparation (filtering)</h2>
<p>Now we can run our standard workflow. But we have to remember to filte our files first. This particular sequencing was of quite low quality, so we have to trim and screen quite harshly.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fastqF &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="st">"_1.fastq.gz"</span>, fastqFiles, <span class="dt">value =</span> <span class="ot">TRUE</span>)
fastqR &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="st">"_2.fastq.gz"</span>, fastqFiles, <span class="dt">value =</span> <span class="ot">TRUE</span>)

samples &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">"_1.fastq</span><span class="ch">\\</span><span class="st">.gz"</span>, <span class="st">"</span><span class="ch">\\</span><span class="st">1"</span>, <span class="kw">basename</span>(fastqF))

filt_path &lt;-<span class="st"> "./filtered"</span>
if(!<span class="kw">file_test</span>(<span class="st">"-d"</span>, filt_path)) <span class="kw">dir.create</span>(filt_path)

filtFs &lt;-<span class="st"> </span><span class="kw">file.path</span>(filt_path, <span class="kw">paste0</span>(samples, <span class="st">"_F_filt.fastq.gz"</span>))
<span class="kw">names</span>(filtFs) &lt;-<span class="st"> </span>samples
filtRs &lt;-<span class="st"> </span><span class="kw">file.path</span>(filt_path, <span class="kw">paste0</span>(samples, <span class="st">"_R_filt.fastq.gz"</span>))
<span class="kw">names</span>(filtRs) &lt;-<span class="st"> </span>samples

filter.track &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">seq_along</span>(fastqF),  function (i) {
    <span class="kw">filterAndTrim</span>(fastqF[i], filtFs[i], fastqR[i], filtRs[i],
                  <span class="dt">truncLen=</span><span class="kw">c</span>(<span class="dv">170</span>,<span class="dv">170</span>), <span class="dt">minLen=</span><span class="kw">c</span>(<span class="dv">170</span>,<span class="dv">170</span>), 
                  <span class="dt">maxN=</span><span class="dv">0</span>, <span class="dt">maxEE=</span><span class="dv">2</span>, <span class="dt">truncQ=</span><span class="dv">2</span>, 
                  <span class="dt">compress=</span><span class="ot">TRUE</span>, <span class="dt">verbose=</span><span class="ot">TRUE</span>)
})

<span class="kw">names</span>(filtFs) &lt;-<span class="st"> </span><span class="kw">names</span>(filtRs) &lt;-<span class="st"> </span>samples

files &lt;-<span class="st"> </span><span class="kw"><a href="../reference/PairedReadFileSet-class.html">PairedReadFileSet</a></span>(filtFs, filtRs)</code></pre></div>
</div>
<div id="data-preparation-primer-set" class="section level2">
<h2 class="hasAnchor">
<a href="#data-preparation-primer-set" class="anchor"></a>Data preparation (primer set)</h2>
<p>Now we have our sequencing read input data, only the primers are still missing. The primer set for the above is included in the MultiAmplicon package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
primer.file &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"real_world_primers.csv"</span>, <span class="dt">package =</span> <span class="st">"MultiAmplicon"</span>)

ptable &lt;-<span class="st"> </span><span class="kw">read.csv</span>(primer.file, <span class="dt">sep=</span><span class="st">","</span>, <span class="dt">header=</span><span class="ot">TRUE</span>, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>)

primerF &lt;-<span class="st"> </span>ptable[, <span class="st">"TS.SequenceF"</span>]
primerR &lt;-<span class="st"> </span>ptable[, <span class="st">"TS.SequenceR"</span>]

<span class="kw">names</span>(primerF) &lt;-<span class="st"> </span><span class="kw">as.character</span>(ptable[, <span class="st">"corrected.NameF"</span>])
<span class="kw">names</span>(primerR) &lt;-<span class="st"> </span><span class="kw">as.character</span>(ptable[, <span class="st">"corrected.NameR"</span>])

primers &lt;-<span class="st"> </span><span class="kw"><a href="../reference/PrimerPairsSet-class.html">PrimerPairsSet</a></span>(primerF, primerR)</code></pre></div>
</div>
<div id="running-the-multiamplicon-pipeline" class="section level2">
<h2 class="hasAnchor">
<a href="#running-the-multiamplicon-pipeline" class="anchor"></a>Running the MultiAmplicon pipeline</h2>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#data-preparation-download">Data preparation (download)</a></li>
      <li><a href="#data-preparation-filtering">Data preparation (filtering)</a></li>
      <li><a href="#data-preparation-primer-set">Data preparation (primer set)</a></li>
      <li><a href="#running-the-multiamplicon-pipeline">Running the MultiAmplicon pipeline</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Emanuel Heitlinger.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
