<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>MultiAmplicon - a small example • MultiAmplicon</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">MultiAmplicon</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/MultiAmplicon-real-world-example.html">MultiAmplicon - a real world example</a>
    </li>
    <li>
      <a href="../articles/MultiAmplicon-small-example.html">MultiAmplicon - a small example</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>MultiAmplicon - a small example</h1>
                        <h4 class="author">Emanuel Heitlinger</h4>
            
            <h4 class="date">2018-04-10</h4>
          </div>

    
    
<div class="contents">
<div id="preparing-sequencing-read-files-and-a-primer-set" class="section level2">
<h2 class="hasAnchor">
<a href="#preparing-sequencing-read-files-and-a-primer-set" class="anchor"></a>Preparing sequencing read files and a primer set</h2>
<p>The MultiAmplicon package uses dedicated classes to ensure that input to its <code>sortAmplicons</code> function is provided correctly. Concerning input files, this means for the user that file names (including paths to the directory they are stored in) of forward and reverse sequencing read files must be provided as a <code>PairedReadFileSet</code>. These files can contain reads in fastq or in gzipped fastq format. We focus as a default use case on (non-interleaved) paired-end reads, as these are the most common format Illumina sequencing reads are currently produced in.</p>
<p>Similarly, primer pairs must be provided in a <code>PrimerPairsSet</code>. We then match forward and reverse primers at the start of forward and reverse reads, respectively. The maximal number of mismatches and the maximal distance from the start of the read can be supplied as arguments to the function. The primer sequences are trimmed from the reads and the read-pair is associated with the amplicon defined by the primer-pair.</p>
<p>We first have a look at how <code>PairedReadFileSet</code> and <code>PrimerPairsSet</code> are generated. Please use files containing quality filtered sequencing reads for the steps below. See the <a href="https://benjjneb.github.io/dada2/tutorial.html">dada2 tutorial</a> or our <a href="https://derele.github.io/MultiAmplicon/articles/MultiAmplicon-real-world-example.html">real-world example vignette</a> for how files are trimmed and quality filtered. But we here first use a very small data set included in the MultiAmplicon package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(MultiAmplicon)

fastq.dir &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"fastq"</span>, <span class="dt">package =</span> <span class="st">"MultiAmplicon"</span>)
fastq.files &lt;-<span class="st"> </span><span class="kw">list.files</span>(fastq.dir, <span class="dt">full.names=</span><span class="ot">TRUE</span>)

Ffastq.file &lt;-<span class="st"> </span>fastq.files[<span class="kw">grepl</span>(<span class="st">"F_filt"</span>, fastq.files)]
Rfastq.file &lt;-<span class="st"> </span>fastq.files[<span class="kw">grepl</span>(<span class="st">"R_filt"</span>, fastq.files)]

PRF &lt;-<span class="st"> </span><span class="kw"><a href="../reference/PairedReadFileSet-class.html">PairedReadFileSet</a></span>(Ffastq.file, Rfastq.file)

primerF &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dt">Amp1F =</span> <span class="st">"AGAGTTTGATCCTGGCTCAG"</span>, <span class="dt">Amp2F =</span> <span class="st">"ACTCCTACGGGAGGCAGC"</span>,
             <span class="dt">Amp3F =</span> <span class="st">"GAATTGACGGAAGGGCACC"</span>, <span class="dt">Amp4F =</span> <span class="st">"YGGTGRTGCATGGCCGYT"</span>,
             <span class="dt">Amp5F =</span> <span class="st">"AAAAACCCCGGGGGGTTTTT"</span>, <span class="dt">Amp6F =</span> <span class="st">"AGAGTTTGATCCTGCCTCAG"</span>)

primerR &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dt">Amp1R =</span> <span class="st">"CTGCWGCCNCCCGTAGG"</span>, <span class="dt">Amp2R =</span> <span class="st">"GACTACHVGGGTATCTAATCC"</span>,
             <span class="dt">Amp3R =</span> <span class="st">"AAGGGCATCACAGACCTGTTAT"</span>, <span class="dt">Amp4R =</span> <span class="st">"TCCTTCTGCAGGTTCACCTAC"</span>,
             <span class="dt">Amp5R =</span> <span class="st">"AAAAACCCCGGGGGGTTTTT"</span>, <span class="dt">Amp6R =</span> <span class="st">"CCTACGGGTGGCAGATGCAG"</span>)

PPS &lt;-<span class="st"> </span><span class="kw"><a href="../reference/PrimerPairsSet-class.html">PrimerPairsSet</a></span>(primerF, primerR)

MA &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MultiAmplicon-class.html">MultiAmplicon</a></span>(PPS, PRF)</code></pre></div>
<p>Note that the primer pairs in the <code>PrimerPairsSet</code> can be named and the same also is true for the sequencing read file pairs (although not in our example above). This results in names for primer-pairs and for samples.</p>
</div>
<div id="sorting-into-multiple-amplicons" class="section level2">
<h2 class="hasAnchor">
<a href="#sorting-into-multiple-amplicons" class="anchor"></a>Sorting into multiple amplicons</h2>
<p>Based on the primer sequences we can now sort the sequences into amplicons. This is done by the function <code>sortAmplicons</code>, which updates the MultiAmplicon object with the sorted amplicons.</p>
<p>For this small example data set we can look at the resulting <code>rawCounts</code> i.e. the number of sequences for each of the fully stratified amplicon x sample matrix.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MA1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sortAmplicons.html">sortAmplicons</a></span>(MA)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">knitr::<span class="kw">kable</span>(<span class="kw"><a href="../reference/MultiAmplicon-class.html">rawCounts</a></span>(MA1))</code></pre></div>
<table class="table">
<thead><tr class="header">
<th></th>
<th align="right">S00_F_filt.fastq.gz</th>
<th align="right">S01_F_filt.fastq.gz</th>
<th align="right">S02_F_filt.fastq.gz</th>
<th align="right">S03_F_filt.fastq.gz</th>
<th align="right">S04_F_filt.fastq.gz</th>
<th align="right">S05_F_filt.fastq.gz</th>
<th align="right">S06_F_filt.fastq.gz</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Amp1F.Amp1R</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">3495</td>
<td align="right">5</td>
<td align="right">5</td>
<td align="right">29310</td>
<td align="right">2569</td>
</tr>
<tr class="even">
<td>Amp2F.Amp2R</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">2949</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">20391</td>
<td align="right">2690</td>
</tr>
<tr class="odd">
<td>Amp3F.Amp3R</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1817</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">2</td>
<td align="right">11671</td>
</tr>
<tr class="even">
<td>Amp4F.Amp4R</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1087</td>
<td align="right">0</td>
<td align="right">7</td>
<td align="right">2076</td>
<td align="right">428</td>
</tr>
<tr class="odd">
<td>Amp5F.Amp5R</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td>Amp6F.Amp6R</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
</div>
<div id="visualizing-read-numbers-per-amplicon-and-filtering-based-on-these" class="section level2">
<h2 class="hasAnchor">
<a href="#visualizing-read-numbers-per-amplicon-and-filtering-based-on-these" class="anchor"></a>Visualizing read numbers per amplicon and filtering based on these</h2>
<p>Now we can plot the result. The function plotAmpliconNumbers allows us also to store clustering information (see <a href="https://www.rdocumentation.org/packages/pheatmap/versions/0.2/topics/pheatmap"></a>). Which can afterwards be used to filter data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clusters &lt;-<span class="st"> </span><span class="kw"><a href="../reference/plotAmpliconNumbers.html">plotAmpliconNumbers</a></span>(MA1)</code></pre></div>
<p><img src="MultiAmplicon-small-example_files/figure-html/plotAmpliconNumbers-1.png" width="672"></p>
<p>We can use this clustering information now to subset our MultiAmplicon object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">two.clusters.row &lt;-<span class="st"> </span><span class="kw">cutree</span>(clusters$tree_row, <span class="dt">k=</span><span class="dv">2</span>)
two.clusters.col &lt;-<span class="st"> </span><span class="kw">cutree</span>(clusters$tree_col, <span class="dt">k=</span><span class="dv">2</span>)

knitr::<span class="kw">kable</span>(two.clusters.row)</code></pre></div>
<table class="table">
<thead><tr class="header">
<th></th>
<th align="right">x</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Amp1F.Amp1R</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td>Amp2F.Amp2R</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td>Amp3F.Amp3R</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td>Amp4F.Amp4R</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td>Amp5F.Amp5R</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td>Amp6F.Amp6R</td>
<td align="right">2</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">knitr::<span class="kw">kable</span>(two.clusters.col)</code></pre></div>
<table class="table">
<thead><tr class="header">
<th></th>
<th align="right">x</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>S00_F_filt.fastq.gz</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td>S01_F_filt.fastq.gz</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td>S02_F_filt.fastq.gz</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td>S03_F_filt.fastq.gz</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td>S04_F_filt.fastq.gz</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td>S05_F_filt.fastq.gz</td>
<td align="right">2</td>
</tr>
<tr class="odd">
<td>S06_F_filt.fastq.gz</td>
<td align="right">2</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
MA.sub &lt;-<span class="st"> </span>MA1[<span class="kw">which</span>(two.clusters.row==<span class="dv">1</span>), <span class="kw">which</span>(two.clusters.col==<span class="dv">2</span>)]

knitr::<span class="kw">kable</span>(<span class="kw"><a href="../reference/MultiAmplicon-class.html">rawCounts</a></span>(MA.sub))</code></pre></div>
<table class="table">
<thead><tr class="header">
<th></th>
<th align="right">S02_F_filt.fastq.gz</th>
<th align="right">S05_F_filt.fastq.gz</th>
<th align="right">S06_F_filt.fastq.gz</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Amp1F.Amp1R</td>
<td align="right">3495</td>
<td align="right">29310</td>
<td align="right">2569</td>
</tr>
<tr class="even">
<td>Amp2F.Amp2R</td>
<td align="right">2949</td>
<td align="right">20391</td>
<td align="right">2690</td>
</tr>
<tr class="odd">
<td>Amp3F.Amp3R</td>
<td align="right">1817</td>
<td align="right">2</td>
<td align="right">11671</td>
</tr>
<tr class="even">
<td>Amp4F.Amp4R</td>
<td align="right">1087</td>
<td align="right">2076</td>
<td align="right">428</td>
</tr>
</tbody>
</table>
<p>In real world applications (see <a href="https://derele.github.io/MultiAmplicon/articles/MultiAmplicon-real-world-example.html">our other vignette</a>) the samples to be excluded from further analysis can be selected based on clustering with negative controls (similar to the example data here where S00 is an empty file and S01 contains one sequence).</p>
</div>
<div id="running-the-amplicon-sequencing-pipeline" class="section level2">
<h2 class="hasAnchor">
<a href="#running-the-amplicon-sequencing-pipeline" class="anchor"></a>Running the amplicon sequencing pipeline</h2>
<p>We first estimate the errors separately for both for dada’s probabilistic sequence variant inference. We do this here on the combined stratified files for different amplicons. This should be done separately for forward ad reverse sequence though!</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">errF &lt;-<span class="st"> </span><span class="kw">learnErrors</span>(<span class="kw">unlist</span>(<span class="kw"><a href="../reference/MultiAmplicon-class.html">stratifiedFilesF</a></span>(MA1)), <span class="dt">nread=</span><span class="fl">1e6</span>,
                    <span class="dt">verbose=</span><span class="dv">0</span>)
errR &lt;-<span class="st"> </span><span class="kw">learnErrors</span>(<span class="kw">unlist</span>(<span class="kw"><a href="../reference/MultiAmplicon-class.html">stratifiedFilesR</a></span>(MA1)), <span class="dt">nread=</span><span class="fl">1e6</span>,
                    <span class="dt">verbose=</span><span class="dv">0</span>)</code></pre></div>
<p>We can now run the pipeline of the dada2 package in parallel over the different amplicons. We update our object from each previous step and safe it under a new name, here for illustration. In your own workflow you can also overwrite the old object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MA2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/derepMulti.html">derepMulti</a></span>(MA.sub, <span class="dt">mc.cores=</span><span class="dv">1</span>) 

MA3 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/dadaMulti.html">dadaMulti</a></span>(MA2, <span class="dt">Ferr=</span>errF, <span class="dt">Rerr=</span>errR,  <span class="dt">pool=</span><span class="ot">FALSE</span>)
<span class="co">#&gt; Sample 1 - 3495 reads in 1493 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 29310 reads in 7224 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 2569 reads in 1053 unique sequences.</span>
<span class="co">#&gt; Sample 1 - 3495 reads in 1535 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 29310 reads in 8273 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 2569 reads in 994 unique sequences.</span>
<span class="co">#&gt; Sample 1 - 2949 reads in 1142 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 20391 reads in 4136 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 2690 reads in 968 unique sequences.</span>
<span class="co">#&gt; Sample 1 - 2949 reads in 1290 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 20391 reads in 5719 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 2690 reads in 1100 unique sequences.</span>
<span class="co">#&gt; Sample 1 - 1817 reads in 589 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 2 reads in 2 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 11671 reads in 3783 unique sequences.</span>
<span class="co">#&gt; Sample 1 - 1817 reads in 547 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 2 reads in 2 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 11671 reads in 4054 unique sequences.</span>
<span class="co">#&gt; Sample 1 - 1087 reads in 394 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 2076 reads in 959 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 428 reads in 166 unique sequences.</span>
<span class="co">#&gt; Sample 1 - 1087 reads in 433 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 2076 reads in 962 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 428 reads in 183 unique sequences.</span>

MA4 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/mergeMulti.html">mergeMulti</a></span>(MA3)

MA5 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sequenceTableMulti.html">sequenceTableMulti</a></span>(MA4)

MA6 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/noChimeMulti.html">noChimeMulti</a></span>(MA5, <span class="dt">mc.cores=</span><span class="dv">1</span>)</code></pre></div>
<p>Ready! This was <em>easy</em>, right?</p>
</div>
<div id="a-word-on-dada2-sequence-inference" class="section level2">
<h2 class="hasAnchor">
<a href="#a-word-on-dada2-sequence-inference" class="anchor"></a>A word on dada2 sequence inference</h2>
<p>The above workflow assumes that errors could be estimated for each amplicon together. This would be beneficial if errors were largely similar between different amplicons (i.e. not influenced by the primer starting the sequence). Below an alternative approach is demonstrated with error probabilities estimated separately for each amplicon (here also overwrite objects from each previous processing steps).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MA.alt &lt;-<span class="st"> </span><span class="kw"><a href="../reference/dadaMulti.html">dadaMulti</a></span>(MA2, <span class="dt">selfConsist=</span><span class="ot">TRUE</span>, <span class="dt">pool=</span><span class="ot">FALSE</span>)
<span class="co">#&gt; Initializing error rates to maximum possible estimate.</span>
<span class="co">#&gt; Sample 1 - 3495 reads in 1493 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 29310 reads in 7224 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 2569 reads in 1053 unique sequences.</span>
<span class="co">#&gt;    selfConsist step 2 </span>
<span class="co">#&gt;    selfConsist step 3 </span>
<span class="co">#&gt;    selfConsist step 4 </span>
<span class="co">#&gt;    selfConsist step 5 </span>
<span class="co">#&gt;    selfConsist step 6 </span>
<span class="co">#&gt; Convergence after  6  rounds.</span>
<span class="co">#&gt; Initializing error rates to maximum possible estimate.</span>
<span class="co">#&gt; Sample 1 - 3495 reads in 1535 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 29310 reads in 8273 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 2569 reads in 994 unique sequences.</span>
<span class="co">#&gt;    selfConsist step 2 </span>
<span class="co">#&gt;    selfConsist step 3 </span>
<span class="co">#&gt;    selfConsist step 4 </span>
<span class="co">#&gt;    selfConsist step 5 </span>
<span class="co">#&gt; Convergence after  5  rounds.</span>
<span class="co">#&gt; Initializing error rates to maximum possible estimate.</span>
<span class="co">#&gt; Sample 1 - 2949 reads in 1142 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 20391 reads in 4136 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 2690 reads in 968 unique sequences.</span>
<span class="co">#&gt;    selfConsist step 2 </span>
<span class="co">#&gt;    selfConsist step 3 </span>
<span class="co">#&gt;    selfConsist step 4 </span>
<span class="co">#&gt; Convergence after  4  rounds.</span>
<span class="co">#&gt; Initializing error rates to maximum possible estimate.</span>
<span class="co">#&gt; Sample 1 - 2949 reads in 1290 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 20391 reads in 5719 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 2690 reads in 1100 unique sequences.</span>
<span class="co">#&gt;    selfConsist step 2 </span>
<span class="co">#&gt;    selfConsist step 3 </span>
<span class="co">#&gt;    selfConsist step 4 </span>
<span class="co">#&gt;    selfConsist step 5 </span>
<span class="co">#&gt; Convergence after  5  rounds.</span>
<span class="co">#&gt; Initializing error rates to maximum possible estimate.</span>
<span class="co">#&gt; Sample 1 - 1817 reads in 589 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 2 reads in 2 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 11671 reads in 3783 unique sequences.</span>
<span class="co">#&gt;    selfConsist step 2 </span>
<span class="co">#&gt;    selfConsist step 3 </span>
<span class="co">#&gt;    selfConsist step 4 </span>
<span class="co">#&gt; Convergence after  4  rounds.</span>
<span class="co">#&gt; Initializing error rates to maximum possible estimate.</span>
<span class="co">#&gt; Sample 1 - 1817 reads in 547 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 2 reads in 2 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 11671 reads in 4054 unique sequences.</span>
<span class="co">#&gt;    selfConsist step 2 </span>
<span class="co">#&gt;    selfConsist step 3 </span>
<span class="co">#&gt;    selfConsist step 4 </span>
<span class="co">#&gt; Convergence after  4  rounds.</span>
<span class="co">#&gt; Initializing error rates to maximum possible estimate.</span>
<span class="co">#&gt; Sample 1 - 1087 reads in 394 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 2076 reads in 959 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 428 reads in 166 unique sequences.</span>
<span class="co">#&gt;    selfConsist step 2 </span>
<span class="co">#&gt;    selfConsist step 3 </span>
<span class="co">#&gt;    selfConsist step 4 </span>
<span class="co">#&gt; Convergence after  4  rounds.</span>
<span class="co">#&gt; Initializing error rates to maximum possible estimate.</span>
<span class="co">#&gt; Sample 1 - 1087 reads in 433 unique sequences.</span>
<span class="co">#&gt; Sample 2 - 2076 reads in 962 unique sequences.</span>
<span class="co">#&gt; Sample 3 - 428 reads in 183 unique sequences.</span>
<span class="co">#&gt;    selfConsist step 2 </span>
<span class="co">#&gt;    selfConsist step 3 </span>
<span class="co">#&gt;    selfConsist step 4 </span>
<span class="co">#&gt; Convergence after  4  rounds.</span>

MA.alt &lt;-<span class="st"> </span><span class="kw"><a href="../reference/mergeMulti.html">mergeMulti</a></span>(MA.alt)

MA.alt &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sequenceTableMulti.html">sequenceTableMulti</a></span>(MA.alt)

MA.alt &lt;-<span class="st"> </span><span class="kw"><a href="../reference/noChimeMulti.html">noChimeMulti</a></span>(MA.alt, <span class="dt">mc.cores=</span><span class="dv">1</span>)</code></pre></div>
<p>I will add a comparison between the two different versions here.</p>
<!-- ```{r compdata, cache=TRUE, message=FALSE} -->
<!-- together <- lapply(MA6@sequenceTableNoChime, function (x) if(all(dim(x)>1))getSequences(x))[[3]] -->
<!-- alt <- lapply(MA.alt@sequenceTableNoChime, function (x) if(all(dim(x)>1))getSequences(x))[[3]] -->
<!-- table(alt%in%together) -->
<!-- alt.diff <- alt[!alt%in%together] -->
<!-- tog.diff <- together[!together%in%alt] -->
<!-- ``` -->
</div>
<div id="using-different-parameters-for-each-amplicon" class="section level2">
<h2 class="hasAnchor">
<a href="#using-different-parameters-for-each-amplicon" class="anchor"></a>Using different parameters for each amplicon</h2>
<p>In some real world applications it might be desirable to have more control over parameters used in the pipeline for different amplicons.</p>
<p>We use ellipsis (aka “…”, dots, dot-dot-dot or three-dots) to pass parameters on to the underlying dada2 functions. The MultiAmplicon package allows all those arguments to be vectors with values varying for different amplicons.</p>
<p>Let’s suppose for this we e.g. want to merge a subset of our amplicons, while we concatenate others using N’s.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">MA.mixed &lt;-<span class="st"> </span><span class="kw"><a href="../reference/mergeMulti.html">mergeMulti</a></span>(MA3, <span class="dt">justConcatenate=</span><span class="kw">c</span>(<span class="ot">TRUE</span>, <span class="ot">FALSE</span>, <span class="ot">FALSE</span>, <span class="ot">TRUE</span>))
<span class="co">#&gt; merging sequences from  3 samples for amplicon  Amp1F.Amp1R </span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; DONE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from  3 samples for amplicon  Amp2F.Amp2R </span>
<span class="co">#&gt; calling mergePairs with justConcatenate=FALSE parameters</span>
<span class="co">#&gt; DONE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from  3 samples for amplicon  Amp3F.Amp3R </span>
<span class="co">#&gt; calling mergePairs with justConcatenate=FALSE parameters</span>
<span class="co">#&gt; DONE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; merging sequences from  3 samples for amplicon  Amp4F.Amp4R </span>
<span class="co">#&gt; calling mergePairs with justConcatenate=TRUE parameters</span>
<span class="co">#&gt; DONE</span></code></pre></div>
<p>The above example would make sense if amplicons Amp1F.Amp1R and Amp4F.Amp4R (1st and 4th in our object) had too little overlap between forward and reverse reads to be merged and would better be concatenated.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#preparing-sequencing-read-files-and-a-primer-set">Preparing sequencing read files and a primer set</a></li>
      <li><a href="#sorting-into-multiple-amplicons">Sorting into multiple amplicons</a></li>
      <li><a href="#visualizing-read-numbers-per-amplicon-and-filtering-based-on-these">Visualizing read numbers per amplicon and filtering based on these</a></li>
      <li><a href="#running-the-amplicon-sequencing-pipeline">Running the amplicon sequencing pipeline</a></li>
      <li><a href="#a-word-on-dada2-sequence-inference">A word on dada2 sequence inference</a></li>
      <li><a href="#using-different-parameters-for-each-amplicon">Using different parameters for each amplicon</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Emanuel Heitlinger.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
