<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>A class combining sequences of forward and reverse primers (in a
<code><a href='PrimerPairsSet-class.html'>PrimerPairsSet-class</a></code>) plus file names of paired end
sequencing files (in a <code><a href='PairedReadFileSet-class.html'>PairedReadFileSet-class</a></code>). — MultiAmplicon-class • MultiAmplicon</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">MultiAmplicon</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/MultiAmplicon-real-world-example.html">MultiAmplicon - a real world example</a>
    </li>
    <li>
      <a href="../articles/MultiAmplicon-small-example.html">MultiAmplicon - a small example</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>A class combining sequences of forward and reverse primers (in a
<code><a href='PrimerPairsSet-class.html'>PrimerPairsSet-class</a></code>) plus file names of paired end
sequencing files (in a <code><a href='PairedReadFileSet-class.html'>PairedReadFileSet-class</a></code>).</h1>
    </div>

    
    <p>The MultiAmplicon class is a container for storing primer pairs,
read files and processed data in an 'amplicon x samples'
format. The object is incrementally filled by running functions
(from the package <code>dada2</code>) with additional slots.</p>
<p>Subsetting for MultiAmplicon objects should conveniently subset
all (potentially) filled slots</p>
    

    <pre class="usage">MultiAmplicon(PrimerPairsSet = PrimerPairsSet(),
  PairedReadFileSet = PairedReadFileSet(), rawCounts = matrix(),
  stratifiedFiles = list(), derep = list(), dada = list(),
  mergers = list(), sequenceTable = list(), sequenceTableNoChime = list(),
  sequenceTableFilled = list())

# S4 method for MultiAmplicon
colnames(x)

# S4 method for MultiAmplicon
rownames(x)

# S4 method for MultiAmplicon
ncol(x)

# S4 method for MultiAmplicon
nrow(x)

# S4 method for MultiAmplicon
dim(x)

rawCounts(x)

# S4 method for MultiAmplicon
rawCounts(x)

getStratifiedFilesF(MA)

getStratifiedFilesR(MA)

getDerepF(MA)

getDerepR(MA)

getDadaF(MA)

getDadaR(MA)

getMergers(MA)

getSequenceTable(MA)

getSequenceTableNoChime(MA)

getSequencesFromTable(MA)

# S4 method for MultiAmplicon
getSequencesFromTable(MA)

# S4 method for MultiAmplicon
show(object)

# S4 method for MultiAmplicon,index,index,ANY
[(x, i, j, ..., drop = FALSE)

# S4 method for MultiAmplicon,index,missing,ANY
[(x, i, j, ..., drop = FALSE)

# S4 method for MultiAmplicon,missing,index,ANY
[(x, i, j, ..., drop = FALSE)

# S4 method for MultiAmplicon,missing,missing,ANY
[(x, i, j, ..., drop = FALSE)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>PrimerPairsSet</th>
      <td><p>a set of primer pairs specifiying your
amplicons see <code><a href='PrimerPairsSet-class.html'>PrimerPairsSet-class</a></code></p></td>
    </tr>
    <tr>
      <th>PairedReadFileSet</th>
      <td><p>a set of paired end sequencing data files
<code><a href='PairedReadFileSet-class.html'>PairedReadFileSet-class</a></code></p></td>
    </tr>
    <tr>
      <th>rawCounts</th>
      <td><p>Users should not supply this parameter, the slot
is created by <code><a href='sortAmplicons.html'>sortAmplicons</a></code>.</p></td>
    </tr>
    <tr>
      <th>stratifiedFiles</th>
      <td><p>Users should not supply this parameter, the
slot is created by <code><a href='sortAmplicons.html'>sortAmplicons</a></code>.</p></td>
    </tr>
    <tr>
      <th>derep</th>
      <td><p>Users should not supply this parameter, the slot is
created by <code><a href='derepMulti.html'>derepMulti</a></code></p></td>
    </tr>
    <tr>
      <th>dada</th>
      <td><p>Users should not supply this parameter, the slot is
created by <code><a href='dadaMulti.html'>dadaMulti</a></code></p></td>
    </tr>
    <tr>
      <th>mergers</th>
      <td><p>Users should not supply this parameter, the slot is
created by <code><a href='mergeMulti.html'>mergeMulti</a></code></p></td>
    </tr>
    <tr>
      <th>sequenceTable</th>
      <td><p>Users should not supply this parameter, the
slot is created by <code><a href='makeSequenceTableMulti.html'>makeSequenceTableMulti</a></code></p></td>
    </tr>
    <tr>
      <th>sequenceTableNoChime</th>
      <td><p>Users should not supply this parameter,
the slot is created by <code><a href='removeChimeraMulti.html'>removeChimeraMulti</a></code></p></td>
    </tr>
    <tr>
      <th>sequenceTableFilled</th>
      <td><p>Users should not supply this parameter,
the slot is created by <code><a href='fillSampleTables.html'>fillSampleTables</a></code></p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>MultiAmplicon-class object</p></td>
    </tr>
    <tr>
      <th>MA</th>
      <td><p>MultiAmplicon-class object</p></td>
    </tr>
    <tr>
      <th>object</th>
      <td><p>A <code>MultiAmplicon-class</code> object.</p></td>
    </tr>
    <tr>
      <th>i</th>
      <td><p>numeric, logical or names vector for subsetting rows (==
amplicons)</p></td>
    </tr>
    <tr>
      <th>j</th>
      <td><p>numeric, logical or names vector for subsetting columns
(== read files, corresponding usually to samples)</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>not used</p></td>
    </tr>
    <tr>
      <th>drop</th>
      <td><p>should not be used</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="functions"><a class="anchor" href="#functions"></a>Functions</h2>

    <ul>
<li><p><code>MultiAmplicon</code>: Constructor for
MultiAmplicon-class</p></li>
</ul>
    
    <h2 class="hasAnchor" id="slots"><a class="anchor" href="#slots"></a>Slots</h2>

    
    <dl class='dl-horizontal'>
<dt><code>PrimerPairsSet</code></dt><dd><p>The primer pairs used in your experiment to
specify amplicons stored in a
<code><a href='PrimerPairsSet-class.html'>PrimerPairsSet-class</a></code> object.</p></dd>
    <dt><code>PairedReadFileSet</code></dt><dd><p>The (quality filtered) fastq files (one
file pair for each sample) that store your sequencing data.</p></dd>
    <dt><code>rawCounts</code></dt><dd><p>A numeric matrix of sequencing read counts per
amplicon and sample. Created by the function
<code><a href='sortAmplicons.html'>sortAmplicons</a></code></p></dd>
    <dt><code>stratifiedFiles</code></dt><dd><p>temporary files as a result of stratifying
into amplicons and samples using the function
<code><a href='sortAmplicons.html'>sortAmplicons</a></code>. Forward (sometimes called R1) and
reverse (sometimes called R2) files are stored as a (amplicons
x samples) matrix of <code><a href='PairedReadFileSet-class.html'>PairedReadFileSet-class</a></code>
objects.</p></dd>
    <dt><code>derep</code></dt><dd><p>A <code><a href='PairedDerep-class.html'>PairedDerep-class</a></code> object containing
pairs of derep-class objects created by <code>dada2</code>’s
<code><a href='http://www.rdocumentation.org/packages/dada2/topics/derepFastq'>derepFastq</a></code> function</p></dd>
    <dt><code>dada</code></dt><dd><p>A <code><a href='PairedDada-class.html'>PairedDada-class</a></code> object containing
pairs of dada-class objects created by <code>dada2</code>’s
<code><a href='http://www.rdocumentation.org/packages/dada2/topics/dada'>dada</a></code> function</p></dd>
    <dt><code>mergers</code></dt><dd><p></p></dd>
    <dt><code>sequenceTable</code></dt><dd><p></p></dd>
    <dt><code>sequenceTableNoChime</code></dt><dd><p></p></dd>
    <dt><code>sequenceTableFilled</code></dt><dd><p>MultiAmplicon(PrimerPairsSet, PairedReadFileSet)</p></dd>
</dl>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p><code><a href='http://www.rdocumentation.org/packages/dada2/topics/derepFastq'>derepFastq</a></code>,<code><a href='http://www.rdocumentation.org/packages/dada2/topics/dada'>dada</a></code></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='no'>primerF</span> <span class='kw'>&lt;-</span> <span class='fu'>c</span>(<span class='st'>"AGAGTTTGATCCTGGCTCAG"</span>, <span class='st'>"ACTCCTACGGGAGGCAGC"</span>,
            <span class='st'>"GAATTGACGGAAGGGCACC"</span>, <span class='st'>"YGGTGRTGCATGGCCGYT"</span>)
<span class='no'>primerR</span> <span class='kw'>&lt;-</span> <span class='fu'>c</span>(<span class='st'>"CTGCWGCCNCCCGTAGG"</span>, <span class='st'>"GACTACHVGGGTATCTAATCC"</span>,
             <span class='st'>"AAGGGCATCACAGACCTGTTAT"</span>, <span class='st'>"TCCTTCTGCAGGTTCACCTAC"</span>)

<span class='no'>PPS</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='PrimerPairsSet-class.html'>PrimerPairsSet</a></span>(<span class='no'>primerF</span>, <span class='no'>primerR</span>)

<span class='no'>fastq.dir</span> <span class='kw'>&lt;-</span> <span class='fu'>system.file</span>(<span class='st'>"extdata"</span>, <span class='st'>"fastq"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"MultiAmplicon"</span>)
<span class='no'>fastq.files</span> <span class='kw'>&lt;-</span> <span class='fu'>list.files</span>(<span class='no'>fastq.dir</span>, <span class='kw'>full.names</span><span class='kw'>=</span><span class='fl'>TRUE</span>)
<span class='no'>Ffastq.file</span> <span class='kw'>&lt;-</span> <span class='no'>fastq.files</span>[<span class='fu'>grepl</span>(<span class='st'>"F_filt"</span>, <span class='no'>fastq.files</span>)]
<span class='no'>Rfastq.file</span> <span class='kw'>&lt;-</span> <span class='no'>fastq.files</span>[<span class='fu'>grepl</span>(<span class='st'>"R_filt"</span>, <span class='no'>fastq.files</span>)]

<span class='no'>PRF</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='PairedReadFileSet-class.html'>PairedReadFileSet</a></span>(<span class='no'>Ffastq.file</span>, <span class='no'>Rfastq.file</span>)

<span class='no'>MA</span> <span class='kw'>&lt;-</span> <span class='fu'>MultiAmplicon</span>(<span class='no'>PPS</span>, <span class='no'>PRF</span>)

<span class='co'>## the number of amplicons (primer pairs)</span>
<span class='fu'>nrow</span>(<span class='no'>MA</span>)</div><div class='output co'>#&gt; [1] 4</div><div class='input'>
<span class='co'>## the number of samples (sequencing read file pairs)</span>
<span class='fu'>ncol</span>(<span class='no'>MA</span>)</div><div class='output co'>#&gt; [1] 7</div><div class='input'>
<span class='co'>## sort into amplicons</span>
<span class='no'>MA1</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='sortAmplicons.html'>sortAmplicons</a></span>(<span class='no'>MA</span>)</div><div class='output co'>#&gt; <span class='message'>using existing directory /tmp/RtmpeEHDm3 and adding specific prefixes to avoid problems when running code twice</span></div><div class='output co'>#&gt; <span class='message'>finished sorting 0 sequencing reads for S00_F_filt.fastq.gz in</span>
#&gt; <span class='message'> /localstorage/ele/git_projects/MultiAmplicon/inst/extdata/fastq/S00_F_filt.fastq.gz and </span>
#&gt; <span class='message'> /localstorage/ele/git_projects/MultiAmplicon/inst/extdata/fastq/S00_R_filt.fastq.gz</span>
#&gt; <span class='message'> into 0 amplicons</span></div><div class='output co'>#&gt; <span class='message'>finished sorting 0 sequencing reads for S01_F_filt.fastq.gz in</span>
#&gt; <span class='message'> /localstorage/ele/git_projects/MultiAmplicon/inst/extdata/fastq/S01_F_filt.fastq.gz and </span>
#&gt; <span class='message'> /localstorage/ele/git_projects/MultiAmplicon/inst/extdata/fastq/S01_R_filt.fastq.gz</span>
#&gt; <span class='message'> into 0 amplicons</span></div><div class='output co'>#&gt; <span class='message'>finished sorting 9348 sequencing reads for S02_F_filt.fastq.gz in</span>
#&gt; <span class='message'> /localstorage/ele/git_projects/MultiAmplicon/inst/extdata/fastq/S02_F_filt.fastq.gz and </span>
#&gt; <span class='message'> /localstorage/ele/git_projects/MultiAmplicon/inst/extdata/fastq/S02_R_filt.fastq.gz</span>
#&gt; <span class='message'> into 4 amplicons</span></div><div class='output co'>#&gt; <span class='message'>finished sorting 5 sequencing reads for S03_F_filt.fastq.gz in</span>
#&gt; <span class='message'> /localstorage/ele/git_projects/MultiAmplicon/inst/extdata/fastq/S03_F_filt.fastq.gz and </span>
#&gt; <span class='message'> /localstorage/ele/git_projects/MultiAmplicon/inst/extdata/fastq/S03_R_filt.fastq.gz</span>
#&gt; <span class='message'> into 1 amplicons</span></div><div class='output co'>#&gt; <span class='message'>finished sorting 14 sequencing reads for S04_F_filt.fastq.gz in</span>
#&gt; <span class='message'> /localstorage/ele/git_projects/MultiAmplicon/inst/extdata/fastq/S04_F_filt.fastq.gz and </span>
#&gt; <span class='message'> /localstorage/ele/git_projects/MultiAmplicon/inst/extdata/fastq/S04_R_filt.fastq.gz</span>
#&gt; <span class='message'> into 3 amplicons</span></div><div class='output co'>#&gt; <span class='message'>finished sorting 51777 sequencing reads for S05_F_filt.fastq.gz in</span>
#&gt; <span class='message'> /localstorage/ele/git_projects/MultiAmplicon/inst/extdata/fastq/S05_F_filt.fastq.gz and </span>
#&gt; <span class='message'> /localstorage/ele/git_projects/MultiAmplicon/inst/extdata/fastq/S05_R_filt.fastq.gz</span>
#&gt; <span class='message'> into 3 amplicons</span></div><div class='output co'>#&gt; <span class='message'>finished sorting 17358 sequencing reads for S06_F_filt.fastq.gz in</span>
#&gt; <span class='message'> /localstorage/ele/git_projects/MultiAmplicon/inst/extdata/fastq/S06_F_filt.fastq.gz and </span>
#&gt; <span class='message'> /localstorage/ele/git_projects/MultiAmplicon/inst/extdata/fastq/S06_R_filt.fastq.gz</span>
#&gt; <span class='message'> into 4 amplicons</span></div><div class='input'>
<span class='no'>MA2</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='derepMulti.html'>derepMulti</a></span>(<span class='no'>MA1</span>)

<span class='no'>MA3</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='dadaMulti.html'>dadaMulti</a></span>(<span class='no'>MA2</span>, <span class='kw'>selfConsist</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>amplicon AGAGTTTGATCCTGGCTCAG.CTGCWGCCNCCCGTAGG: dada estimation of sequence variants from 5 of 7 possible sample files</span></div><div class='output co'>#&gt; <span class='message'>calling dada with selfConsist=TRUE parameters</span></div><div class='output co'>#&gt; Initializing error rates to maximum possible estimate.
#&gt; Sample 1 - 3495 reads in 1493 unique sequences.
#&gt; Sample 2 - 5 reads in 5 unique sequences.
#&gt; Sample 3 - 5 reads in 5 unique sequences.
#&gt; Sample 4 - 29310 reads in 7224 unique sequences.
#&gt; Sample 5 - 2569 reads in 1053 unique sequences.
#&gt;    selfConsist step 2 
#&gt;    selfConsist step 3 
#&gt;    selfConsist step 4 
#&gt;    selfConsist step 5 
#&gt; Convergence after  5  rounds.
#&gt; Initializing error rates to maximum possible estimate.
#&gt; Sample 1 - 3495 reads in 1535 unique sequences.
#&gt; Sample 2 - 5 reads in 5 unique sequences.
#&gt; Sample 3 - 5 reads in 4 unique sequences.
#&gt; Sample 4 - 29310 reads in 8273 unique sequences.
#&gt; Sample 5 - 2569 reads in 994 unique sequences.
#&gt;    selfConsist step 2 
#&gt;    selfConsist step 3 
#&gt;    selfConsist step 4 
#&gt;    selfConsist step 5 
#&gt; Convergence after  5  rounds.</div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>amplicon ACTCCTACGGGAGGCAGC.GACTACHVGGGTATCTAATCC: dada estimation of sequence variants from 4 of 7 possible sample files</span></div><div class='output co'>#&gt; <span class='message'>calling dada with selfConsist=TRUE parameters</span></div><div class='output co'>#&gt; Initializing error rates to maximum possible estimate.
#&gt; Sample 1 - 2949 reads in 1142 unique sequences.
#&gt; Sample 2 - 2 reads in 2 unique sequences.
#&gt; Sample 3 - 20391 reads in 4136 unique sequences.
#&gt; Sample 4 - 2690 reads in 968 unique sequences.
#&gt;    selfConsist step 2 
#&gt;    selfConsist step 3 
#&gt;    selfConsist step 4 
#&gt; Convergence after  4  rounds.
#&gt; Initializing error rates to maximum possible estimate.
#&gt; Sample 1 - 2949 reads in 1290 unique sequences.
#&gt; Sample 2 - 2 reads in 2 unique sequences.
#&gt; Sample 3 - 20391 reads in 5719 unique sequences.
#&gt; Sample 4 - 2690 reads in 1100 unique sequences.
#&gt;    selfConsist step 2 
#&gt;    selfConsist step 3 
#&gt;    selfConsist step 4 
#&gt;    selfConsist step 5 
#&gt; Convergence after  5  rounds.</div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>amplicon GAATTGACGGAAGGGCACC.AAGGGCATCACAGACCTGTTAT: dada estimation of sequence variants from 2 of 7 possible sample files</span></div><div class='output co'>#&gt; <span class='message'>calling dada with selfConsist=TRUE parameters</span></div><div class='output co'>#&gt; Initializing error rates to maximum possible estimate.
#&gt; Sample 1 - 1817 reads in 589 unique sequences.
#&gt; Sample 2 - 11671 reads in 3783 unique sequences.
#&gt;    selfConsist step 2 
#&gt;    selfConsist step 3 
#&gt;    selfConsist step 4 
#&gt; Convergence after  4  rounds.
#&gt; Initializing error rates to maximum possible estimate.
#&gt; Sample 1 - 1817 reads in 547 unique sequences.
#&gt; Sample 2 - 11671 reads in 4054 unique sequences.
#&gt;    selfConsist step 2 
#&gt;    selfConsist step 3 
#&gt;    selfConsist step 4 
#&gt; Convergence after  4  rounds.</div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>amplicon YGGTGRTGCATGGCCGYT.TCCTTCTGCAGGTTCACCTAC: dada estimation of sequence variants from 4 of 7 possible sample files</span></div><div class='output co'>#&gt; <span class='message'>calling dada with selfConsist=TRUE parameters</span></div><div class='output co'>#&gt; Initializing error rates to maximum possible estimate.
#&gt; Sample 1 - 1087 reads in 394 unique sequences.
#&gt; Sample 2 - 7 reads in 6 unique sequences.
#&gt; Sample 3 - 2076 reads in 959 unique sequences.
#&gt; Sample 4 - 428 reads in 166 unique sequences.
#&gt;    selfConsist step 2 
#&gt;    selfConsist step 3 
#&gt;    selfConsist step 4 
#&gt; Convergence after  4  rounds.
#&gt; Initializing error rates to maximum possible estimate.
#&gt; Sample 1 - 1087 reads in 433 unique sequences.
#&gt; Sample 2 - 7 reads in 5 unique sequences.
#&gt; Sample 3 - 2076 reads in 962 unique sequences.
#&gt; Sample 4 - 428 reads in 183 unique sequences.
#&gt;    selfConsist step 2 
#&gt;    selfConsist step 3 
#&gt;    selfConsist step 4 
#&gt; Convergence after  4  rounds.</div><div class='input'>
<span class='no'>MA4</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='mergeMulti.html'>mergeMulti</a></span>(<span class='no'>MA3</span>, <span class='kw'>justConcatenate</span><span class='kw'>=</span><span class='fl'>TRUE</span>)</div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>merging sequences from 5 samples for amplicon AGAGTTTGATCCTGGCTCAG.CTGCWGCCNCCCGTAGG</span></div><div class='output co'>#&gt; <span class='message'>calling mergePairs with justConcatenate=TRUE parameters</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>merging sequences from 4 samples for amplicon ACTCCTACGGGAGGCAGC.GACTACHVGGGTATCTAATCC</span></div><div class='output co'>#&gt; <span class='message'>calling mergePairs with justConcatenate=TRUE parameters</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>merging sequences from 2 samples for amplicon GAATTGACGGAAGGGCACC.AAGGGCATCACAGACCTGTTAT</span></div><div class='output co'>#&gt; <span class='message'>calling mergePairs with justConcatenate=TRUE parameters</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>merging sequences from 4 samples for amplicon YGGTGRTGCATGGCCGYT.TCCTTCTGCAGGTTCACCTAC</span></div><div class='output co'>#&gt; <span class='message'>calling mergePairs with justConcatenate=TRUE parameters</span></div><div class='input'>
<span class='no'>MA5</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='makeSequenceTableMulti.html'>makeSequenceTableMulti</a></span>(<span class='no'>MA4</span>)</div><div class='output co'>#&gt; <span class='message'>calling makeSequenceTable with default parameters</span></div><div class='output co'>#&gt; <span class='message'>calling makeSequenceTable with default parameters</span></div><div class='output co'>#&gt; <span class='message'>calling makeSequenceTable with default parameters</span></div><div class='output co'>#&gt; <span class='message'>calling makeSequenceTable with default parameters</span></div><div class='input'>
<span class='no'>MA6</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='removeChimeraMulti.html'>removeChimeraMulti</a></span>(<span class='no'>MA5</span>, <span class='kw'>mc.cores</span><span class='kw'>=</span><span class='fl'>1</span>)</div><div class='output co'>#&gt; <span class='message'>calling removeBimeraDenovo with default parameters</span></div><div class='output co'>#&gt; <span class='message'>As of the 1.4 release, the default method changed to consensus (from pooled).</span></div><div class='output co'>#&gt; <span class='message'>calling removeBimeraDenovo with default parameters</span></div><div class='output co'>#&gt; <span class='message'>As of the 1.4 release, the default method changed to consensus (from pooled).</span></div><div class='output co'>#&gt; <span class='message'>calling removeBimeraDenovo with default parameters</span></div><div class='output co'>#&gt; <span class='message'>As of the 1.4 release, the default method changed to consensus (from pooled).</span></div><div class='output co'>#&gt; <span class='message'>calling removeBimeraDenovo with default parameters</span></div><div class='output co'>#&gt; <span class='message'>As of the 1.4 release, the default method changed to consensus (from pooled).</span></div><div class='input'>
<span class='no'>MA7</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='fillSampleTables.html'>fillSampleTables</a></span>(<span class='no'>MA6</span>)</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#functions">Functions</a></li>

      <li><a href="#slots">Slots</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Emanuel Heitlinger

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Emanuel Heitlinger.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
